core/pathresolver.go

```diff
@@ -29,29 +38,30 @@ func resolveRecurse(n *IpfsNode, p path.Path, depth int) (*merkledag.Node, error
 	// emerges when resolving just a <hash>. Is it meant
 	// to be an ipfs or an ipns resolution?
 
-	if strings.HasPrefix(p.String(), "/ipns/") {
+	if strings.HasPrefix(r.p.String(), "/ipns/") {
 		// if it's an ipns path, try to resolve it.
 		// if we can't, we can give that error back to the user.
-		seg := p.Segments()
+		seg := r.p.Segments()
 		if len(seg) < 2 || seg[1] == "" { // just "/ipns/"
-			return nil, fmt.Errorf("invalid path: %s", string(p))
+			return nil, fmt.Errorf("invalid path: %s", string(r.p))
 		}
 
 		ipnsPath := seg[1]
 		extensions := seg[2:]
-		respath, err := n.Namesys.Resolve(n.Context(), ipnsPath)
+		respath, err := r.n.Namesys.Resolve(r.ctx, ipnsPath)
 		if err != nil {
 			return nil, err
 		}
 
 		segments := append(respath.Segments(), extensions...)
-		respath, err = path.FromSegments(segments...)
+		r.p, err = path.FromSegments(segments...)
 		if err != nil {
 			return nil, err
 		}
-		return resolveRecurse(n, respath, depth+1)
+		return r.resolveRecurse(depth + 1)
 	}
 
 	// ok, we have an ipfs path now (or what we'll treat as one)
-	return n.Resolver.ResolvePath(p)
+	// TODO(cryptix): we are loosing the context from the initial Resolve(ctx, ...) call here
```

Comment:
Thought so, too.

Also turned up some uses of `context.TODO()` which make more sense now.